### PR TITLE
Stats page

### PR DESCRIPTION
**Pull request by lfalvarez**
This adds a very basic stats page per instance.

### CODE REVIEW

File: nuntium/templates/nuntium/profiles/stats.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{% extends "base_edit.html" %}
+{% load i18n %}
+{% load pagination_tags %}
+{% load nuntium_tags %}
+{% load staticfiles %}
+
+{% block extrajs %}
```
Comment (duncanparkes) [inline nuntium/templates/nuntium/profiles/stats.html]: There's no need to include the block if you're not putting anything in it and not overriding anything.

File: nuntium/templates/nuntium/profiles/stats.html [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{% endblock extrajs %}
+
+{% block extracss %}
+<link href="https://gitcdn.github.io/bootstrap-toggle/2.2.0/css/bootstrap-toggle.min.css" rel="stylesheet">
+{% endblock extracss %}
+
+{% block extrascripts %}
```
Comment (duncanparkes) [inline nuntium/templates/nuntium/profiles/stats.html]: What's the difference between blocks 'extrajs' and 'extrascripts'? Is it just two different places you can add javascript in? I wonder if there is a better name for them if so.

File: nuntium/user_section/tests/stats_page_tests.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    def test_there_is_a_url(self):
+        '''There is a url for stats per instance'''
+        url = reverse('stats', kwargs={'pk': self.writeitinstance.pk})
+        self.assertTrue(url)
+
+    def test_get_the_url_brings_the_instance_and_renders_the_template(self):
+        '''When getting the url brings some stats and renders the specific template'''
+        url = reverse('stats', kwargs={'pk': self.writeitinstance.pk})
+        client = Client()
```
Comment (duncanparkes) [inline nuntium/user_section/tests/stats_page_tests.py]: The Django TestCase should provide you with `self.client`, so there's no need to import it or instantiate it for each test.

File: nuntium/user_section/tests/stats_page_tests.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        self.assertIn('writeitinstance', response.context)
+        self.assertTemplateUsed(response, 'nuntium/profiles/stats.html')
+
+    def test_it_brings_some_stats_as_well(self):
+        '''It brings some stats as well'''
+        url = reverse('stats', kwargs={'pk': self.writeitinstance.pk})
+        client = Client()
+        client.login(username=self.writeitinstance.owner, password="feroz")
+        response = client.get(url)
+        self.assertIn('stats', response.context)
```
Comment (duncanparkes) [inline nuntium/user_section/tests/stats_page_tests.py]: This is likely to be true of any page which links to the correct one as well.

File: nuntium/user_section/tests/stats_page_tests.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        self.assertIsInstance(stats, StatsPerInstance)
+        self.assertEquals(stats.writeitinstance, self.writeitinstance)
+
+
+class StatsPerInstanceTestCase(TestCase):
+    def setUp(self):
+        super(StatsPerInstanceTestCase, self).setUp()
+        self.writeitinstance = WriteItInstance.objects.get(id=1)
+
+    def test_instanciate(self):
```
Comment (duncanparkes) [inline nuntium/user_section/tests/stats_page_tests.py]: s/instanciate/instantiate/

File: nuntium/user_section/views.py [hunk 6]
```diff
@@ -414,3 +414,50 @@ class WriteItDeleteView(WriteItInstanceOwnerMixin, DeleteView):
...
+            ('public_messages_with_answers', self.public_messages_with_answers),
+        ]
+
+    @property
+    def amount_of_messages(self):
+        return self.writeitinstance.message_set.count()
+
+    @property
+    def amount_of_public_messages(self):
+        return self.writeitinstance.message_set.filter(public=True).count()
```
Comment (duncanparkes) [inline nuntium/user_section/views.py]: Do you want all public messages, or just the ones that have been confirmed?

File: nuntium/user_section/views.py [hunk 7]
```diff
@@ -414,3 +414,50 @@ class WriteItDeleteView(WriteItInstanceOwnerMixin, DeleteView):
...
+    def amount_of_private_messages(self):
+        return self.writeitinstance.message_set.filter(public=False).count()
+
+    @property
+    def public_messages_with_answers(self):
+        # I could perhaps try to use
+        # a Count aggregation instead of doing this
+        # for loop
+        # https://docs.djangoproject.com/en/1.7/topics/db/aggregation/
+        result = 0
```
Comment (duncanparkes) [inline nuntium/user_section/views.py]: The aggregate version might look something like:

```
return (self.writeitinstance.message_set
    .annotate(num_answers=Count('answers'))
    .filter(public=True, num_answers__gt=0)
    .count())
```

File: nuntium/user_section/tests/stats_page_tests.py (hunk 5 above)
Comment (lfalvarez) [inline nuntium/user_section/tests/stats_page_tests.py]: Uh sorry

File: nuntium/user_section/views.py (hunk 6 above)
Comment (lfalvarez) [inline nuntium/user_section/views.py]: Good question. I thought of all public messages, but I'm open to comments here

File: nuntium/user_section/views.py (hunk 7 above)
Comment (lfalvarez) [inline nuntium/user_section/views.py]: Cool adding it in!

File: nuntium/templates/nuntium/profiles/stats.html (hunk 2 above)
Comment (lfalvarez) [inline nuntium/templates/nuntium/profiles/stats.html]: You're right but I think I should open a new issue for it as it is a broader thing than just this issue. What do you say?

File: nuntium/templates/nuntium/profiles/stats.html (hunk 1 above)
Comment (lfalvarez) [inline nuntium/templates/nuntium/profiles/stats.html]: Yeah that's right but I'm wondering if to fix that here, I rather create a different issue =)

**Comment by lfalvarez** [comment]
I think that this one is pretty much ready. I'm gonna merge it.

File: nuntium/user_section/views.py (hunk 6 above)
Comment (tmtmtmtm) [inline nuntium/user_section/views.py]: Both?

File: nuntium/templates/nuntium/profiles/stats.html (hunk 2 above)
Comment (tmtmtmtm) [inline nuntium/templates/nuntium/profiles/stats.html]: Sure — just open it as a new issue.

File: nuntium/user_section/views.py (hunk 6 above)
Comment (lfalvarez) [inline nuntium/user_section/views.py]: good! I think it is worth adding a new method for getting all the confirmed public messages.

Comment (lfalvarez) [inline nuntium/user_section/views.py]: https://github.com/ciudadanointeligente/write-it/commit/30f4214efcaa8f3e78800f8955cdf6f312a0d45b adds a Public and Confirmed messages.

**Pull request closed without merging**